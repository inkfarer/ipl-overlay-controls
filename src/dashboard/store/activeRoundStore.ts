import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
