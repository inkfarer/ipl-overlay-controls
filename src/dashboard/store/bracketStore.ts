import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
