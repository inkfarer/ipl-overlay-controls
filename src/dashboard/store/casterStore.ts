import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const useCasterStore = defineStore('casters', {
    state: () => ({
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    } as CasterStore),
    actions: {
        updateCaster({ id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        removeUncommittedCaster(id: string): void {
            delete this.uncommittedCasters[id];
        },
        addUncommittedCaster({ id, caster }: { id: string, caster?: Caster }): void {
            this.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        addDefaultCaster(): string {
            const id = generateId();
            this.addUncommittedCaster({ id });
            return id;
        },
        async saveUncommittedCaster({ id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            this.removeUncommittedCaster(id);
            return newId;
        },
        async loadCastersFromVc(): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    this.addUncommittedCaster({ id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        async removeCaster(id: string): Promise<void> {
            return nodecg.sendMessage('removeCaster', { id });
        }
    }
});
