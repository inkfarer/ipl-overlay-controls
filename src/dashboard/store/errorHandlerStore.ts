import { App } from 'vue';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    }),
    actions: {
        removeRecentError({ key }: { key: string }): void {
            delete this.recentErrors[key];
        },
        handleError({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(this.recentErrors).length >= 2) return;

            const id = generateId();
            this.recentErrors[id] = err;
            window.setTimeout(() => {
                delete this.recentErrors[id];
            }, 25000);
        }
    }
});

export function setUpErrorHandler(app: App<unknown>): void {
    const store = useErrorHandlerStore();
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info });
    };
}
