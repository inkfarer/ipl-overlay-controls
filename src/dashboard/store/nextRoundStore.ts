import { NextRound } from 'schemas';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { BeginNextMatchRequest } from 'types/messages/activeRound';
import { defineStore } from 'pinia';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const useNextRoundStore = defineStore('nextRound', {
    state: () => ({
        nextRound: null
    } as NextRoundStore),
    actions: {
        setShowOnStream(newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        },
        beginNextMatch(request: BeginNextMatchRequest): void {
            nodecg.sendMessage('beginNextMatch', request);
        },
        setNextRound(request: SetNextRoundRequest): void {
            nodecg.sendMessage('setNextRound', request);
        }
    }
});
