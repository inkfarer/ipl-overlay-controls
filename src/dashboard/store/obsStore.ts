import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetObsDataRequest } from 'types/messages/obs';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async connect(store, { address, password }: { address: string, password?: string }): Promise<void> {
            return nodecg.sendMessage('connectToObs', { address, password });
        },
        async setData(store, data: SetObsDataRequest): Promise<void> {
            return nodecg.sendMessage('setObsData', data);
        },
        async startGame(): Promise<void> {
            return nodecg.sendMessage('startGame');
        },
        async endGame(): Promise<void> {
            return nodecg.sendMessage('endGame');
        },
        setEnabled(store, enabled: boolean): void {
            nodecg.sendMessage('setObsSocketEnabled', enabled);
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
