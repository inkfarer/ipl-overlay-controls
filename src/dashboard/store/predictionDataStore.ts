import { NodeCGBrowser } from 'nodecg/browser';
import { Prediction, PredictionStore } from 'schemas';
import { PredictionStatus } from 'types/enums/predictionStatus';
import { defineStore } from 'pinia';

const predictionStore = nodecg.Replicant<PredictionStore>('predictionStore');

export const predictionReps = [ predictionStore ];

export interface PredictionDataStore {
    predictionStore: PredictionStore;
}

export const usePredictionDataStore = defineStore('prediction', {
    state: () => ({
        predictionStore: null,
    } as PredictionDataStore),
    actions: {
        async lockPrediction() {
            return patchPrediction(
                this.predictionStore.currentPrediction,
                PredictionStatus.LOCKED,
                [ PredictionStatus.ACTIVE ]);
        },
        async cancelPrediction() {
            return patchPrediction(
                this.predictionStore.currentPrediction,
                PredictionStatus.CANCELED,
                [ PredictionStatus.ACTIVE, PredictionStatus.LOCKED ]);
        },
        async resolvePrediction({ winningOutcomeIndex }: { winningOutcomeIndex: number }) {
            if (winningOutcomeIndex > 1 || winningOutcomeIndex < 0) {
                throw new Error(`Cannot resolve prediction with outcome index ${winningOutcomeIndex}.`);
            } else if (!this.predictionStore.currentPrediction?.id) {
                throw new Error('No prediction to resolve.');
            } else if (this.predictionStore.currentPrediction.status !== PredictionStatus.LOCKED ) {
                throw new Error('Can only resolve a locked prediction.');
            }

            return nodecg.sendMessage('patchPrediction', {
                id: this.predictionStore.currentPrediction.id,
                status: PredictionStatus.RESOLVED,
                winning_outcome_id: this.predictionStore.currentPrediction.outcomes[winningOutcomeIndex].id
            });
        },
        async createPrediction(
            { title, teamAName, teamBName, duration }:
                { title: string, teamAName: string, teamBName: string, duration: number }
        ) {
            if (
                [PredictionStatus.ACTIVE, PredictionStatus.LOCKED]
                    .includes(this.predictionStore.currentPrediction?.status as PredictionStatus)
            ) {
                throw new Error('An unresolved prediction already exists.');
            }

            return nodecg.sendMessage('postPrediction', {
                title,
                outcomes: [
                    { title: teamAName },
                    { title: teamBName }
                ],
                prediction_window: duration
            });
        },
        async reconnect() {
            return nodecg.sendMessage('reconnectToRadiaSocket');
        }
    }
});

async function patchPrediction(
    currentPrediction: Prediction,
    status: PredictionStatus,
    allowedForStatuses: PredictionStatus[]
): Promise<void> {
    if (!currentPrediction || !allowedForStatuses.includes(currentPrediction?.status as PredictionStatus)) {
        throw new Error(`Cannot set prediction status at this time. Status must match one of [${allowedForStatuses.join(', ')}]`);
    }

    return nodecg.sendMessage('patchPrediction', {
        id: currentPrediction.id,
        status: status,
    });
}
