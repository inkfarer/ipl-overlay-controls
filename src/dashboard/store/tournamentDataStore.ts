import { MatchStore, RoundStore, TournamentData } from 'schemas';
import {
    GetTournamentDataRequest,
    GetTournamentDataResponse,
    ToggleTeamImageRequest
} from 'types/messages/tournamentData';
import { UpdateRoundStoreRequest, UpdateRoundStoreResponse } from 'types/messages/roundStore';
import { defineStore } from 'pinia';
import i18next from 'i18next';

const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');
const roundStore = nodecg.Replicant<RoundStore>('roundStore');
const matchStore = nodecg.Replicant<MatchStore>('matchStore');

export const tournamentDataReps = [ tournamentData, roundStore, matchStore ];

export interface TournamentDataStore {
    tournamentData: TournamentData;
    roundStore: RoundStore;
    matchStore: MatchStore;
}

export const useTournamentDataStore = defineStore('tournamentData', {
    state: () => ({
        tournamentData: null,
        roundStore: null,
        matchStore: null
    } as TournamentDataStore),
    actions: {
        async getTournamentData({ method, id }: GetTournamentDataRequest): Promise<GetTournamentDataResponse> {
            return nodecg.sendMessage('getTournamentData', { method, id });
        },
        async getSmashggEvent({ eventId }: { eventId: number }): Promise<void> {
            return await nodecg.sendMessage('getSmashggEvent', { eventId });
        },
        async uploadTeamData({ file }: { file: File }): Promise<void> {
            return sendLocalFile('teams', file);
        },
        async uploadRoundData({ file }: { file: File }): Promise<void> {
            return sendLocalFile('rounds', file);
        },
        async fetchRoundData({ url }: { url: string }): Promise<void> {
            return nodecg.sendMessage('getRounds', { url });
        },
        async updateRound(data: UpdateRoundStoreRequest): Promise<void> {
            return nodecg.sendMessage('updateRound', data);
        },
        async insertRound(data: UpdateRoundStoreRequest): Promise<UpdateRoundStoreResponse> {
            return nodecg.sendMessage('insertRound', data);
        },
        resetRoundStore() {
            nodecg.sendMessage('resetRoundStore');
        },
        setTeamImageHidden({ teamId, isVisible }: { teamId: string, isVisible: boolean }): Promise<void> {
            return nodecg.sendMessage('toggleTeamImage', { teamId, isVisible } as ToggleTeamImageRequest);
        },
        removeRound({ roundId }: { roundId: string }): Promise<void> {
            return nodecg.sendMessage('removeRound', { roundId });
        },
        setShortName(newValue: string): void {
            tournamentData.value.meta.shortName = newValue;
        },
        async refreshTournamentData(): Promise<void> {
            return nodecg.sendMessage('refreshTournamentData');
        }
    }
});

async function sendLocalFile(dataType: 'teams' | 'rounds', file: File): Promise<void> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('jsonType', dataType);

    const response = await fetch('/ipl-overlay-controls/upload-tournament-json', { method: 'POST', body: formData });

    if (response.status !== 200) {
        const responseText = await response.text();
        throw new Error(i18next.t(
            'common:tournamentDataStore.fileUploadError',
            { statusCode: response.status, message: responseText }));
    }
}
