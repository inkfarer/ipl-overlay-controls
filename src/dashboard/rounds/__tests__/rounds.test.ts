import Rounds from '../rounds.vue';
import { config, mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { TournamentDataStore, tournamentDataStoreKey } from '../../store/tournamentDataStore';
import { nextRoundStoreKey } from '../../store/nextRoundStore';
import { activeRoundStoreKey } from '../../store/activeRoundStore';

describe('Rounds', () => {
    config.global.stubs = {
        FontAwesomeIcon: true,
        IplButton: true,
        RoundEditor: true,
        IplErrorDisplay: true
    };

    const mockResetRoundStore = jest.fn();

    function createTournamentDataStore() {
        return createStore<TournamentDataStore>({
            state: {
                tournamentData: {
                    meta: { id: '1093478', source: 'SMASHGG' },
                    teams: [
                        { id: '123123', name: 'cool team A', players: [], showLogo: true },
                        { id: '345345', name: 'cool team B', players: [], showLogo: false }
                    ]
                },
                roundStore: {
                    '0387': {
                        meta: { name: 'dope round', isCompleted: false },
                        games: []
                    },
                    '9573': {
                        meta: { name: 'dope round the second', isCompleted: false },
                        games: []
                    },
                    '2426': {
                        meta: { name: 'dope round the third', isCompleted: false },
                        games: []
                    }
                }
            },
            actions: {
                resetRoundStore: mockResetRoundStore
            }
        });
    }

    function createActiveRoundStore() {
        return createStore({
            state: {
                activeRound: {
                    round: {
                        id: '9573'
                    }
                }
            }
        });
    }

    function createNextRoundStore() {
        return createStore({
            state: {
                nextRound: {
                    round: {
                        id: '2426'
                    }
                }
            }
        });
    }

    it('matches snapshot', () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot when round sidebar is open', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });

        await wrapper.get('[data-test="open-all-rounds-sidebar"]').trigger('click');

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('resets round store on reset button click', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });

        await wrapper.get('[data-test="open-all-rounds-sidebar"]').trigger('click');
        wrapper.getComponent('[data-test="reset-rounds-button"]').vm.$emit('click');

        expect(mockResetRoundStore).toHaveBeenCalled();
    });

    it('selects expected round by default', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        expect(roundEditor.props().roundId).toEqual('2426');
        expect(roundEditor.props().round).toEqual({
            games: [],
            meta: {
                isCompleted: false,
                name: 'dope round the third'
            }
        });
        expect(roundEditor.props().isNewRound).toEqual(false);
    });

    it('switches selected round when round is selected from sidebar', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        await wrapper.get('[data-test="open-all-rounds-sidebar"]').trigger('click');
        wrapper.getComponent('[data-test="round-option-9573"]').vm.$emit('click');
        await wrapper.vm.$nextTick();

        expect(roundEditor.props().roundId).toEqual('9573');
        expect(roundEditor.props().round).toEqual({
            games: [],
            meta: {
                isCompleted: false,
                name: 'dope round the second'
            }
        });
        expect(roundEditor.props().isNewRound).toEqual(false);
    });

    it('handles new 3-game round being created', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        wrapper.getComponent('[data-test="new-3-game-round"]').vm.$emit('click');
        await wrapper.vm.$nextTick();

        expect(roundEditor.props().round).toEqual({
            games: [
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' }
            ],
            meta: {
                name: 'New Round'
            }
        });
        expect(roundEditor.props().isNewRound).toEqual(true);
    });

    it('handles new 5-game round being created', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        wrapper.getComponent('[data-test="new-5-game-round"]').vm.$emit('click');
        await wrapper.vm.$nextTick();

        expect(roundEditor.props().round).toEqual({
            games: [
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' }
            ],
            meta: {
                name: 'New Round'
            }
        });
        expect(roundEditor.props().isNewRound).toEqual(true);
    });

    it('handles new 7-game round being created', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        wrapper.getComponent('[data-test="new-7-game-round"]').vm.$emit('click');
        await wrapper.vm.$nextTick();

        expect(roundEditor.props().round).toEqual({
            games: [
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' },
                { mode: 'Unknown Mode', stage: 'Unknown Stage' }
            ],
            meta: {
                name: 'New Round'
            }
        });
        expect(roundEditor.props().isNewRound).toEqual(true);
    });

    it('handles round creation being cancelled', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        wrapper.getComponent('[data-test="new-7-game-round"]').vm.$emit('click');
        await wrapper.vm.$nextTick();
        roundEditor.vm.$emit('cancelNewRound');
        await wrapper.vm.$nextTick();

        expect(roundEditor.props().isNewRound).toEqual(false);
    });

    it('handles round being created', async () => {
        const store = createTournamentDataStore();
        const nextRoundStore = createNextRoundStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(Rounds, {
            global: {
                plugins: [
                    [store, tournamentDataStoreKey],
                    [nextRoundStore, nextRoundStoreKey],
                    [activeRoundStore, activeRoundStoreKey]
                ]
            }
        });
        const roundEditor = wrapper.getComponent('[data-test="round-editor"]');

        wrapper.getComponent('[data-test="new-7-game-round"]').vm.$emit('click');
        await wrapper.vm.$nextTick();
        roundEditor.vm.$emit('createNewRound', '0387');
        await wrapper.vm.$nextTick();

        expect(roundEditor.props().isNewRound).toEqual(false);
        expect(roundEditor.props().roundId).toEqual('0387');
        expect(roundEditor.props().round).toEqual({
            meta: { name: 'dope round', isCompleted: false },
            games: []
        });
    });
});
