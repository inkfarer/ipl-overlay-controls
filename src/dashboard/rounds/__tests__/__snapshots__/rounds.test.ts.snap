// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rounds matches snapshot 1`] = `
"<ipl-error-display-stub class=\\"m-b-8\\"></ipl-error-display-stub>
<!---->
<div class=\\"ipl-expansion-panel__content\\" data-v-40cedb04=\\"\\">
  <div class=\\"ipl-expansion-panel__header layout horizontal center-vertical\\" data-v-40cedb04=\\"\\">
    <div class=\\"ipl-expansion-panel__header-background\\" data-v-40cedb04=\\"\\"></div>
    <div class=\\"ipl-expansion-panel__title\\" data-v-40cedb04=\\"\\">Create Round</div>
    <font-awesome-icon-stub icon=\\"chevron-left\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon\\" data-v-40cedb04=\\"\\"></font-awesome-icon-stub>
    <div class=\\"header-extra\\" data-v-40cedb04=\\"\\"></div>
  </div>
  <div class=\\"content\\" data-v-40cedb04=\\"\\" style=\\"display: none;\\">
    <div class=\\"layout horizontal\\">
      <ipl-button-stub label=\\"3 Games\\" color=\\"green\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" data-test=\\"new-3-game-round\\"></ipl-button-stub>
      <ipl-button-stub label=\\"5 Games\\" color=\\"green\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" class=\\"m-l-6\\" data-test=\\"new-5-game-round\\"></ipl-button-stub>
    </div>
    <ipl-button-stub label=\\"7 Games\\" color=\\"green\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" class=\\"m-t-6\\" data-test=\\"new-7-game-round\\"></ipl-button-stub>
  </div>
</div>
<div class=\\"ipl-space color-dark clickable layout horizontal center-vertical round-menu-space m-t-8\\" data-test=\\"open-all-rounds-sidebar\\" data-v-11cc5f1c=\\"\\">
  <font-awesome-icon-stub icon=\\"bars\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"large-icon m-r-8\\"></font-awesome-icon-stub> All Rounds
</div>
<round-editor-stub roundid=\\"2426\\" round=\\"[object Object]\\" isnewround=\\"false\\" class=\\"m-t-8\\" data-test=\\"round-editor\\"></round-editor-stub>"
`;

exports[`Rounds matches snapshot when round sidebar is open 1`] = `
"<ipl-error-display-stub class=\\"m-b-8\\"></ipl-error-display-stub>
<div class=\\"ipl-sidebar__wrapper sidebar-transition-enter-from sidebar-transition-enter-active\\" data-v-2b6fb510=\\"\\">
  <div class=\\"background\\" data-v-2b6fb510=\\"\\"></div>
  <div class=\\"content\\" data-v-2b6fb510=\\"\\">
    <div class=\\"ipl-space color-light\\" data-v-11cc5f1c=\\"\\">
      <ipl-button-stub label=\\"Reset rounds\\" color=\\"red\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"true\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" data-test=\\"reset-rounds-button\\"></ipl-button-stub>
    </div>
    <div class=\\"ipl-space color-light layout horizontal m-t-8\\" data-v-11cc5f1c=\\"\\">
      <div class=\\"color-key color-key-next\\"></div> Next round
    </div>
    <div class=\\"ipl-space color-blue clickable m-t-8 round-option is-next-round\\" data-test=\\"round-option-2426\\" data-v-11cc5f1c=\\"\\">dope round the third</div>
    <div class=\\"ipl-space color-light clickable m-t-8 round-option\\" data-test=\\"round-option-9573\\" data-v-11cc5f1c=\\"\\">dope round the second</div>
    <div class=\\"ipl-space color-light clickable m-t-8 round-option\\" data-test=\\"round-option-0387\\" data-v-11cc5f1c=\\"\\">dope round</div>
  </div>
</div>
<div class=\\"ipl-expansion-panel__content\\" data-v-40cedb04=\\"\\">
  <div class=\\"ipl-expansion-panel__header layout horizontal center-vertical\\" data-v-40cedb04=\\"\\">
    <div class=\\"ipl-expansion-panel__header-background\\" data-v-40cedb04=\\"\\"></div>
    <div class=\\"ipl-expansion-panel__title\\" data-v-40cedb04=\\"\\">Create Round</div>
    <font-awesome-icon-stub icon=\\"chevron-left\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon\\" data-v-40cedb04=\\"\\"></font-awesome-icon-stub>
    <div class=\\"header-extra\\" data-v-40cedb04=\\"\\"></div>
  </div>
  <div class=\\"content\\" data-v-40cedb04=\\"\\" style=\\"display: none;\\">
    <div class=\\"layout horizontal\\">
      <ipl-button-stub label=\\"3 Games\\" color=\\"green\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" data-test=\\"new-3-game-round\\"></ipl-button-stub>
      <ipl-button-stub label=\\"5 Games\\" color=\\"green\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" class=\\"m-l-6\\" data-test=\\"new-5-game-round\\"></ipl-button-stub>
    </div>
    <ipl-button-stub label=\\"7 Games\\" color=\\"green\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" href=\\"\\" class=\\"m-t-6\\" data-test=\\"new-7-game-round\\"></ipl-button-stub>
  </div>
</div>
<div class=\\"ipl-space color-dark clickable layout horizontal center-vertical round-menu-space m-t-8\\" data-test=\\"open-all-rounds-sidebar\\" data-v-11cc5f1c=\\"\\">
  <font-awesome-icon-stub icon=\\"bars\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"large-icon m-r-8\\"></font-awesome-icon-stub> All Rounds
</div>
<round-editor-stub roundid=\\"2426\\" round=\\"[object Object]\\" isnewround=\\"false\\" class=\\"m-t-8\\" data-test=\\"round-editor\\"></round-editor-stub>"
`;
